resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+    compile-time error (_wrong number of type arguments supplied_).
+
+  - Otherwise, `F` must have at least one formal type parameter, and there must
+    be zero type arguments `T` (_in other words, type arguments will be
+    inferred_). Then:
+
+    - Initialize `C` to an empty list of type constraints.
+
+    - Initialize `{U_1, U_2, ...}` to a list of type schemas, of the same length
+      as `{X_1, X_2, ...}`, each of which is `_`.
+
+    - Using subtype constraint generation, attempt to match `R_F` as a subtype
+      of `K` with respect to the list of type variables `{X_1, X_2, ...}`.
+
+      - If this succeeds, then accumulate the resulting list of type constraints
+        into `C`. Then, let `{V_1, V_2, ...}` be the constraint solution for the
+        set of type variables `{X_1, X_2, ...}` with respect to the constaint
+        set `C`, with partial solution `{U_1, U_2, ...}`. Replace `{U_1, U_2,
+        ...}` with `{V_1, V_2, ...}`. _This step is known as "downwards
+        inference", since it has the effect of passing type information __down__
+        the syntax tree from the context of the invocation to the context of the
+        arguments._
+
+      - Otherwise, leave `C` and `{U_1, U_2, ...}` unchanged.
+
+- Partition the arguments `{e_1, e_2, ...}` into stages (see [argument
+  partitioning](#Argument-partitioning) below), and then for each stage _k_:
+
+  - For each `e_i` in stage _k_, let `K_i` be the result of substituting `{U_1,
+    U_2, ...}` for `{X_1, X_2, ...}` in `P_i`.
+
+  - For each `e_i` in stage _k_ that _does not_ take the form of a
+    _&lt;functionExpression&gt;_ enclosed in zero or more parentheses, in order
+    of increasing _i_:
+
+    - Let `m_i_preliminary` be the result of performing expression inference on
+      `e_i`, in context `K_i`.
+
+  - For each `e_i` in stage _k_ that _does_ take the form of a
+    _&lt;functionExpression&gt;_ enclosed in zero or more parentheses, in order
+    of increasing _i_:
+
+    - Let `m_i_preliminary` be the result of performing expression inference on
+      `e_i`, in context `K_i`.
+
+  - _Note that a property of argument partitioning is that arguments that are
+    not function literal expressions are always placed in stage zero, so this
+    has the effect that all arguments that are not function expressions are type
+    inferred first, in the order in which they appear in source code, followed
+    by all arguments that __are__ function expressions, possibly in a staged
+    fashion._
+
+  - If `F` has at least one formal type parameter, and there are zero type
+    arguments `T` (_in other words, generic type inference is being performed_),
+    then:
+
+    - For each `e_i` in stage _k_:
+
+      - Let `S_i` be the static type of `m_i_preliminary`.
+
+      - Using subtype constraint generation, attempt to match `S_i` as a subtype
+        of `K_i` with respect to the list of type variables `{X_1, X_2, ...}`.
```

Comment:
I think this should be `P_i` right?  `K_i` is has been closed WRT the `X_i` by substituting in the partial solutions `U_i` above?

Reply:
Yes, thank you. I double checked the implementation and you are absolutely right. Fixed.

Also I noticed that the “For each `e_i` in stage _k_” bullet enclosing this bullet doesn’t specify the order in which the `e_i` are considered. I’ve clarified that the order is the same as the order in which expression inference was performed.